File: 12899.py
import sys;input=sys.stdin.readline;
M = 2000001
N = int(input())
tree_length = 1
while tree_length < M:
    tree_length <<= 1
tree_length<<=1
segment_tree = [0 for _ in range(tree_length)]
def update(tree_idx, query_start, query_end, array_idx):
    if query_start==query_end:
        segment_tree[tree_idx] += 1
        return 

    mid = (query_start+query_end)//2
    segment_tree[tree_idx] += 1
    if mid >= array_idx:
        update(tree_idx<<1, query_start, mid, array_idx)
    else:
        update((tree_idx<<1)+1, mid+1, query_end, array_idx)

def remove(tree_idx, query_start, query_end, n):
    global number
    if query_start==query_end:
        segment_tree[tree_idx] -= 1
        number = query_start
        return

    mid = (query_start+query_end)//2
    segment_tree[tree_idx] -= 1
    left = tree_idx<<1
    if segment_tree[left] >= n:
        remove(left, query_start, mid, n)
    else:
        remove(left+1, mid+1, query_end, n-segment_tree[left])
    return

for _ in range(N):
    T, X = map(int, input().split())

    if T == 1: # X 추가
        update(1, 0, M-1, X)
    else: # X번째로 작은 수 삭제
        number = 0
        remove(1, 0, M-1, X)
        print(number)

File: Baekjoon/12899.cpp
#include <stdio.h>
int segment_tree[4194304] = {0, };

void update(int treeIdx, int queryStart, int queryEnd, int arrayIdx) {
    if (queryStart==queryEnd) {
        segment_tree[treeIdx] += 1;
        return;
    }

    int mid = (queryStart+queryEnd)>>1;
    segment_tree[treeIdx] += 1;

    if (mid >= arrayIdx)
        update(treeIdx<<1, queryStart, mid, arrayIdx);
    else
        update((treeIdx<<1)+1, mid+1, queryEnd, arrayIdx);
    return;
}

int remove(int treeIdx, int queryStart, int queryEnd, int n) {
    if (queryStart==queryEnd) {
        segment_tree[treeIdx] -= 1;
        return queryStart;
    }

    int mid = (queryStart+queryEnd)>>1;
    segment_tree[treeIdx] -= 1;
    int left = treeIdx<<1;

    if (segment_tree[left] >= n)
        return remove(left, queryStart, mid, n);
    else
        return remove(left+1, mid+1, queryEnd, n-segment_tree[left]);
}

int main() {
    int M = 2000001;
    int N, T, X;
    scanf("%d", &N);

    for (int i = 0; i < N; i++) {
        scanf("%d %d", &T, &X);

        if (T==1) {
            update(1, 0, M-1, X);
        } else {
            printf("%d\n", remove(1, 0, M-1, X));
        }
    }
}